design/extensions.md

```diff
@@ -0,0 +1,313 @@
+* Pulls all necessary extension packages from the registry
+* Unpacks them and installs all owned/defined CRDs and transforms the unpackaged content into an `Extension` CRD instance that serves as a record of the install
+* Starts up the custom controller so that it is in the running state
+* Marks the `ExtensionRequest` status as succeeded
+
+## `ExtensionRequest` CRD
+
+To commence the installation of new functionality into a Crossplane cluster, an instance of the `ExtensionRequest` CRD should be created.
+The EM will be watching for events on this type and it will begin the process of installing an extension during its reconcile loop.
+
+`ExtensionRequests` can be specified by either a package name or by a CRD type.
+When given a CRD type, the controller will query the registry to find out what package owns that CRD and then it will download that package to proceed with the install.
+This gives more flexibility to how extensions are installed and does not require the requestor to know what package a CRD is defined in.
+
+```yaml
+# request to extend Crossplane with the redis package,
+# using a specific version number
+apiVersion: extensions.crossplane.io/v1alpha1
+kind: ExtensionRequest
+metadata:
+  name: redis-from-package
+spec:
+  source: registry.crossplane.io
+  package: redis:v0.1.0
+status:
+  conditions:
+  - type: Ready
+    status: "True"
+---
+# request to extend Crossplane with the package that defines/owns,
+# the rediscluster CRD
+apiVersion: extensions.crossplane.io/v1alpha1
+kind: ExtensionRequest
+metadata:
+  name: redis-from-crd
+spec:
+  source: registry.crossplane.io
+  crd: redisclusters.cache.crossplane.io/v1alpha1
+status:
+  conditions:
+  - type: Creating
+    status: "True"
+```
+
+## `Extension` CRD
+
+The `Extension` CRD serves as a record of an installed extension (a custom controller and its CRDs).
+These records make it so that a user or system can query Crossplane and learn all of the functionality that has been installed on it as well as their statuses.
+
+Instances of this CRD can be generated from the filesystem based contents of a package, i.e. the metadata files contained inside the package.
+This can be thought of as a translation operation, where the file based content is translated into a YAML based version that is stored in the `Extension` CRD.
+
+`Extension` CRD instances can also be created directly by a user without any knowledge of packages at all.
+They can directly create any CRDs that their extension requires and then create an `Extension` CRD instance that describes their extension, its custom controller, etc.
+The Extension Manager will see this new instance and take the steps necessary to ensure the custom controller is running in the cluster and the extension’s functionality is available.
+
+```yaml
+apiVersion: extensions.crossplane.io/v1alpha1
+kind: Extension
```

Comment:
`Extension` feels close enough to the status of an `ExtensionRequest` that I question whether they should really be two different resources. I feels like the main driver for them being separate is that we may want users to to able to "manually" install something, then tell Crossplane it was installed?

Reply:
correct, a user should be able to directly create an instance of the `Extension` CRD, without any need for the package downloading/extracting/etc that is done by `ExtensionRequests`.  The controller watching for the `Extension` type would then create the correct permissions and the start up the right controller so that the entity captured in the `Extension` instance is ready to start serving requests.